File: frontend/src/types/flow.ts
/**
 * フローグラフの型定義
 *
 * 将来 Rust パーサー(WASM)の explain() が返す JSON の契約。
 * 現状はモックデータ(src/mock/flowGraph.ts)がこの型を満たす。
 * docs/06_api_design.md / docs/07_ui_design.md を参照。
 */

/** 列の系譜(lineage)における役割 */
export type ColumnRole =
  /** 最終結果に値が届く列(集計関数経由を含む) */
  | 'output'
  /** 結合キー・WHERE / GROUP BY などの条件でのみ使われる列 */
  | 'used'

export type Column = {
  name: string
  role: ColumnRole
}

export type JoinType = 'INNER' | 'LEFT' | 'RIGHT' | 'FULL' | 'CROSS'

/** グラフのノード。kind ごとに描画コンポーネントが変わる */
export type FlowNode =
  /** FROM で参照される実テーブル */
  | {
      id: string
      kind: 'scan'
      label: string
      alias?: string
      columns: Column[]
      /** SQL に現れていない列が存在しうる場合 true(列リスト末尾に … を描く) */
      hasMore: boolean
      groupId?: string
    }
  /** WITH で定義された共通テーブル式 */
  | {
      id: string
      kind: 'cte'
      label: string
      alias?: string
      columns: Column[]
      hasMore: boolean
      groupId?: string
    }
  /** JOIN の合流結果(JOIN はノードではなく合流として描く) */
  | {
      id: string
      kind: 'joined'
      label: string
      joinType: JoinType
      columns: Column[]
      hasMore: boolean
      groupId?: string
    }
  /** WHERE / HAVING による絞り込み */
  | {
      id: string
      kind: 'filter'
      phase: 'where' | 'having'
      predicate: string
      groupId?: string
    }
  /** GROUP BY によるグループ化 */
  | { id: string; kind: 'group'; keys: string[]; groupId?: string }
  /** 途中の SELECT(サブクエリ・CTE 内の射影) */
  | {
      id: string
      kind: 'project'
      items: string[]
      distinct: boolean
      groupId?: string
    }
  /** ORDER BY による並び替え */
  | { id: string; kind: 'sort'; keys: string[]; groupId?: string }
  /** LIMIT / OFFSET による切り出し */
  | {
      id: string
      kind: 'slice'
      limit?: number
      offset?: number
      groupId?: string
    }
  /** 最終結果のテーブル */
  | {
      id: string
      kind: 'result'
      columns: Column[]
      hasMore: boolean
      groupId?: string
    }

export type FlowNodeKind = FlowNode['kind']

/** ノード間のエッジ。label は JOIN の結合キーなど矢印上に表示する文字列 */
export type FlowEdgeData = {
  source: string
  target: string
  label?: string
}

/** WITH 句などノードをまとめる枠 */
export type FlowGroup = {
  id: string
  label: string
}

/** 論理実行順タイムラインの 1 ステップ */
export type TimelineStep = {
  order: number
  label: string
  nodeIds: string[]
}

export type FlowGraph = {
  nodes: FlowNode[]
  edges: FlowEdgeData[]
  groups: FlowGroup[]
  timeline: TimelineStep[]
}

File: explain/src/flow.rs
//! フロントエンドに渡すフローグラフの型定義
//! frontend/src/types/flow.ts の `FlowGraph` と同じ構造に JSON 化される契約

use serde::Serialize;

/// 列の系譜(lineage)における役割
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum Role {
    /// 最終結果に値が届く列(集計関数経由を含む)
    Output,
    /// 結合キー・WHERE / GROUP BY などの条件でのみ使われる列
    Used,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct Column {
    pub name: String,
    pub role: Role,
}

/// グラフのノード。kind ごとにフロントエンドの描画コンポーネントが変わる
#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(tag = "kind", rename_all = "lowercase")]
pub enum FlowNode {
    /// FROM で参照される実テーブル
    Scan {
        id: String,
        label: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        alias: Option<String>,
        columns: Vec<Column>,
        #[serde(rename = "hasMore")]
        has_more: bool,
        #[serde(rename = "groupId", skip_serializing_if = "Option::is_none")]
        group_id: Option<String>,
    },
    /// WITH で定義された共通テーブル式(FROM 内サブクエリの結果もこれ)
    Cte {
        id: String,
        label: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        alias: Option<String>,
        columns: Vec<Column>,
        #[serde(rename = "hasMore")]
        has_more: bool,
        #[serde(rename = "groupId", skip_serializing_if = "Option::is_none")]
        group_id: Option<String>,
    },
    /// JOIN の合流結果(JOIN はノードではなく合流として描く)
    Joined {
        id: String,
        label: String,
        #[serde(rename = "joinType")]
        join_type: String,
        columns: Vec<Column>,
        #[serde(rename = "hasMore")]
        has_more: bool,
        #[serde(rename = "groupId", skip_serializing_if = "Option::is_none")]
        group_id: Option<String>,
    },
    /// WHERE / HAVING による絞り込み
    Filter {
        id: String,
        phase: String,
        predicate: String,
        #[serde(rename = "groupId", skip_serializing_if = "Option::is_none")]
        group_id: Option<String>,
    },
    /// GROUP BY によるグループ化
    Group {
        id: String,
        keys: Vec<String>,
        #[serde(rename = "groupId", skip_serializing_if = "Option::is_none")]
        group_id: Option<String>,
    },
    /// 途中の SELECT(CTE・サブクエリ内の射影)
    Project {
        id: String,
        items: Vec<String>,
        distinct: bool,
        #[serde(rename = "groupId", skip_serializing_if = "Option::is_none")]
        group_id: Option<String>,
    },
    /// ORDER BY による並び替え
    Sort {
        id: String,
        keys: Vec<String>,
        #[serde(rename = "groupId", skip_serializing_if = "Option::is_none")]
        group_id: Option<String>,
    },
    /// LIMIT / OFFSET による切り出し
    Slice {
        id: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        limit: Option<u64>,
        #[serde(skip_serializing_if = "Option::is_none")]
        offset: Option<u64>,
        #[serde(rename = "groupId", skip_serializing_if = "Option::is_none")]
        group_id: Option<String>,
    },
    /// 最終結果のテーブル
    Result {
        id: String,
        columns: Vec<Column>,
        #[serde(rename = "hasMore")]
        has_more: bool,
        #[serde(rename = "groupId", skip_serializing_if = "Option::is_none")]
        group_id: Option<String>,
    },
}

impl FlowNode {
    pub fn id(&self) -> &str {
        match self {
            FlowNode::Scan { id, .. }
            | FlowNode::Cte { id, .. }
            | FlowNode::Joined { id, .. }
            | FlowNode::Filter { id, .. }
            | FlowNode::Group { id, .. }
            | FlowNode::Project { id, .. }
            | FlowNode::Sort { id, .. }
            | FlowNode::Slice { id, .. }
            | FlowNode::Result { id, .. } => id,
        }
    }
}

/// ノード間のエッジ。label は JOIN の結合キーなど矢印上に表示する文字列
#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct FlowEdge {
    pub source: String,
    pub target: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub label: Option<String>,
}

/// WITH 句などノードをまとめる枠
#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct FlowGroup {
    pub id: String,
    pub label: String,
}

/// 論理実行順タイムラインの 1 ステップ
#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct TimelineStep {
    pub order: u32,
    pub label: String,
    #[serde(rename = "nodeIds")]
    pub node_ids: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct FlowGraph {
    pub nodes: Vec<FlowNode>,
    pub edges: Vec<FlowEdge>,
    pub groups: Vec<FlowGroup>,
    pub timeline: Vec<TimelineStep>,
}
